File: clients/mobile/ios/WidgetReviewKit/Sources/WidgetReviewKit/WidgetReviewSnapshot.swift
//
//  WidgetReviewSnapshot.swift
//  WidgetReviewKit
//
//  The Swift mirror of the Dart `WidgetReviewSnapshot` wire contract
//  (shared/app-core/lib/src/review/widget_review_snapshot.dart). The Dart app builds a snapshot,
//  encodes it (toJson), and writes it into the App Group; the iOS widget decodes it HERE. The two
//  sides are pinned to one committed golden fixture
//  (shared/app-core/test/fixtures/widget_review_snapshot.golden.json) so they can't drift.
//
//  Foundation-only (no WidgetKit / SwiftUI) so the model layer is testable with `swift test`. The
//  widget NEVER computes FSRS — it renders these resolved cards and emits rating events; scheduling
//  stays server-authoritative.
//

import Foundation

// MARK: - Meaning status

/// Whether a card's back (meaning) is renderable. Mirrors the Dart `WidgetMeaningStatus`.
public enum WidgetMeaningStatus: String, Equatable {
  case ready
  case unsupported   // the target language is off the explanation allowlist
  case unavailable   // a fetch failure / no blob — the front is still reviewable

  /// Lenient decode: an unknown / absent wire value degrades to `.unavailable` (the safe "no meaning,
  /// still reviewable" state), so a future status string can't break an older widget — matching the
  /// Dart `WidgetMeaningStatus.fromWire`.
  public static func from(wire: String?) -> WidgetMeaningStatus {
    guard let wire, let status = WidgetMeaningStatus(rawValue: wire) else { return .unavailable }
    return status
  }
}

// MARK: - Target span

/// A half-open UTF-16 `[start, end)` highlight range within a card's context text. On the wire it is a
/// 2-element array `[start, end]` (or null). UTF-16 so Swift's `String.Index`/`NSRange` and Dart's
/// UTF-16 offsets agree. Mirrors the Dart `WidgetTargetSpan`.
public struct WidgetTargetSpan: Equatable {
  public let start: Int
  public let end: Int

  public init(start: Int, end: Int) {
    self.start = start
    self.end = end
  }

  /// Build from the wire `[start, end]` array, or nil if absent / malformed (wrong length / negative /
  /// inverted) — the widget then renders plain text. Mirrors the Dart `WidgetTargetSpan.fromBounds` /
  /// `fromJson`, so an invalid span resolves to nil on BOTH sides (no asymmetric mismatch).
  public static func from(array: [Int]?) -> WidgetTargetSpan? {
    guard let array, array.count == 2 else { return nil }
    let start = array[0], end = array[1]
    guard start >= 0, end >= start else { return nil }
    return WidgetTargetSpan(start: start, end: end)
  }

  /// The substring of [text] this span highlights, or nil if the span is out of range for [text]
  /// (defensive — a stale snapshot's span against a different string). Uses UTF-16 offsets.
  public func substring(of text: String) -> String? {
    let utf16 = text.utf16
    guard start <= end, end <= utf16.count,
          let from = utf16.index(utf16.startIndex, offsetBy: start, limitedBy: utf16.endIndex),
          let to = utf16.index(utf16.startIndex, offsetBy: end, limitedBy: utf16.endIndex),
          let sFrom = String.Index(from, within: text),
          let sTo = String.Index(to, within: text)
    else { return nil }
    return String(text[sFrom..<sTo])
  }
}

// MARK: - Card

/// One fully-resolved card the widget renders: the front (context sentence + the target span to
/// highlight — the word is SHOWN, recognition review, not a cloze blank) and the back ([meaning] +
/// [meaningStatus]). Mirrors the Dart `WidgetReviewCard`.
public struct WidgetReviewCard: Equatable, Decodable {
  public let wordId: String
  public let surfaceUnit: String
  public let targetLang: String
  public let dueAt: Int64           // glance count only — the widget never re-sorts/filters by it
  public let state: String          // "due" | "new" (both reviewable; renderer doesn't branch on it)
  public let contextText: String    // the most-recent sentence, or "" for a bare card
  public let targetSpan: WidgetTargetSpan?
  public let ipa: String?           // bare IPA (primary reading US, else UK), or nil when omitted
  public let meaning: String?       // per-POS senses, one line per part of speech, or nil when not `ready`
  public let meaningStatus: WidgetMeaningStatus
  public let contextMeaning: String?  // in-sentence "Explain here" gloss for the front sentence, or nil

  public init(
    wordId: String, surfaceUnit: String, targetLang: String, dueAt: Int64, state: String,
    contextText: String, targetSpan: WidgetTargetSpan?, ipa: String? = nil,
    meaning: String?, meaningStatus: WidgetMeaningStatus, contextMeaning: String? = nil
  ) {
    self.wordId = wordId
    self.surfaceUnit = surfaceUnit
    self.targetLang = targetLang
    self.dueAt = dueAt
    self.state = state
    self.contextText = contextText
    self.targetSpan = targetSpan
    self.ipa = ipa
    self.meaning = meaning
    self.meaningStatus = meaningStatus
    self.contextMeaning = contextMeaning
  }

  private enum CodingKeys: String, CodingKey {
    case wordId, surfaceUnit, targetLang, dueAt, state, contextText, targetSpan, ipa, meaning,
      meaningStatus, contextMeaning
  }

  public init(from decoder: Decoder) throws {
    let c = try decoder.container(keyedBy: CodingKeys.self)
    wordId = try c.decode(String.self, forKey: .wordId)
    surfaceUnit = try c.decode(String.self, forKey: .surfaceUnit)
    targetLang = try c.decode(String.self, forKey: .targetLang)
    dueAt = try c.decode(Int64.self, forKey: .dueAt)
    state = try c.decodeIfPresent(String.self, forKey: .state) ?? "due"
    contextText = try c.decodeIfPresent(String.self, forKey: .contextText) ?? ""
    targetSpan = WidgetTargetSpan.from(array: try c.decodeIfPresent([Int].self, forKey: .targetSpan))
    ipa = try c.decodeIfPresent(String.self, forKey: .ipa)
    meaning = try c.decodeIfPresent(String.self, forKey: .meaning)
    meaningStatus = WidgetMeaningStatus.from(
      wire: try c.decodeIfPresent(String.self, forKey: .meaningStatus))
    contextMeaning = try c.decodeIfPresent(String.self, forKey: .contextMeaning)
  }
}

// MARK: - Snapshot

/// The pre-resolved review queue the Dart app hands the widget via the App Group. Mirrors the Dart
/// `WidgetReviewSnapshot`. The widget renders [cards] from [cursor] and emits rating events back
/// through the shared offline queue — it never fetches and never computes FSRS.
public struct WidgetReviewSnapshot: Equatable, Decodable {
  /// The newest INCOMPATIBLE wire version this build can read. The Swift decoder is the ENFORCING gate
  /// (the Dart producer is forward-lenient): a snapshot stamped newer than this is REJECTED rather than
  /// mis-rendered (see `init(from:)`).
  public static let currentSchemaVersion = 1

  public let schemaVersion: Int
  public let snapshotId: String
  public let builtAt: Int64
  public let staleAfterMs: Int64
  public let cursor: Int
  public let cards: [WidgetReviewCard]

  public init(
    schemaVersion: Int = WidgetReviewSnapshot.currentSchemaVersion, snapshotId: String,
    builtAt: Int64, staleAfterMs: Int64, cursor: Int, cards: [WidgetReviewCard]
  ) {
    self.schemaVersion = schemaVersion
    self.snapshotId = snapshotId
    self.builtAt = builtAt
    self.staleAfterMs = staleAfterMs
    self.cursor = cursor
    self.cards = cards
  }

  private enum CodingKeys: String, CodingKey {
    case schemaVersion, snapshotId, builtAt, staleAfterMs, cursor, cards
  }

  public init(from decoder: Decoder) throws {
    let c = try decoder.container(keyedBy: CodingKeys.self)
    let version = try c.decodeIfPresent(Int.self, forKey: .schemaVersion)
      ?? WidgetReviewSnapshot.currentSchemaVersion
    // The latch: reject a snapshot written by a NEWER app than this widget can read, rather than
    // silently mis-render it (the Dart-side doc designates Swift as the enforcing gate).
    if version > WidgetReviewSnapshot.currentSchemaVersion {
      throw DecodingError.dataCorrupted(
        .init(
          codingPath: decoder.codingPath,
          debugDescription:
            "WidgetReviewSnapshot schemaVersion \(version) > supported \(WidgetReviewSnapshot.currentSchemaVersion)"
        ))
    }
    schemaVersion = version
    snapshotId = try c.decode(String.self, forKey: .snapshotId)
    builtAt = try c.decode(Int64.self, forKey: .builtAt)
    staleAfterMs = try c.decode(Int64.self, forKey: .staleAfterMs)
    cursor = try c.decodeIfPresent(Int.self, forKey: .cursor) ?? 0
    cards = try c.decodeIfPresent([WidgetReviewCard].self, forKey: .cards) ?? []
  }

  /// Whether the snapshot has aged past its freshness window (mirrors the Dart `isStaleAt`). Past it
  /// the widget shows a glance + "open to refresh" and refuses to grade offline.
  public func isStale(atMillis nowMs: Int64) -> Bool { nowMs - builtAt >= staleAfterMs }

  /// Decode from raw App Group bytes, returning nil on ANY failure (malformed JSON, a rejected
  /// schemaVersion, a missing required field) so the widget degrades to "open to refresh" instead of
  /// crashing the timeline.
  public static func decode(from data: Data) -> WidgetReviewSnapshot? {
    try? JSONDecoder().decode(WidgetReviewSnapshot.self, from: data)
  }
}

File: shared/app-core/lib/src/review/widget_review_snapshot.dart
import 'review_resolve.dart';

/// The pre-resolved review queue the app hands the widget (App → App Group → SwiftUI). The widget
/// NEVER fetches and NEVER computes FSRS — it renders these cards and emits rating events that drain
/// back through the shared offline queue. Pure serialization (no Flutter / no API deps); the SwiftUI
/// side decodes the SAME JSON, pinned by a committed golden fixture so the Dart encoder and the Swift
/// decoder can't drift.
///
/// Schema (widget RFC §5.1): a [snapshotId] (cursor scope + rebuild-reconcile anchor), a [builtAt] +
/// [staleAfterMs] freshness window (a stale snapshot degrades to glance and forbids offline grading),
/// a [cursor], and the [cards]. The queue carries DUE cards then today's server-surfaced NEW cards —
/// the SAME set the in-app review shows — so the widget reviews new words too (each [WidgetReviewCard]
/// records its own `state`). Grading a new card offline can, only when the flush crosses local midnight,
/// shift that card's "introduced today" tally to the next day (the new-card cap is server-receive-time
/// attributed) — a soft-pacing off-by-one, never a scheduling/data error.
class WidgetReviewSnapshot {
  const WidgetReviewSnapshot({
    required this.snapshotId,
    required this.builtAt,
    required this.staleAfterMs,
    required this.cursor,
    required this.cards,
    this.schemaVersion = currentSchemaVersion,
  });

  /// Bumped only on an INCOMPATIBLE wire change. The CONSUMER is the SwiftUI widget (App → App Group →
  /// Swift), so the Swift decoder is the enforcing gate: it rejects a snapshot whose [schemaVersion]
  /// exceeds the one it was built for (rather than mis-render). The Dart side here is the PRODUCER and
  /// is intentionally forward-lenient on decode (app→app round-trips only) — see [fromJson].
  static const int currentSchemaVersion = 1;

  final int schemaVersion;

  /// Identifies this build. The widget's cursor is SCOPED to it (see [resumeCursor]); a rebuild mints
  /// a new id so a stale cursor can't apply to a different card list.
  final String snapshotId;

  /// When the snapshot was built (epoch ms).
  final int builtAt;

  /// How long the snapshot stays fresh (ms). Past it the widget shows a glance + "open to refresh" and
  /// refuses to grade offline (a grade against a stale queue risks a wrong card).
  final int staleAfterMs;

  /// The card the widget should show next (index into [cards]).
  final int cursor;

  final List<WidgetReviewCard> cards;

  /// True when [nowMs] is at or past the freshness window.
  bool isStaleAt(int nowMs) => nowMs - builtAt >= staleAfterMs;

  Map<String, dynamic> toJson() => {
    'schemaVersion': schemaVersion,
    'snapshotId': snapshotId,
    'builtAt': builtAt,
    'staleAfterMs': staleAfterMs,
    'cursor': cursor,
    'cards': [for (final c in cards) c.toJson()],
  };

  factory WidgetReviewSnapshot.fromJson(Map<String, dynamic> json) => WidgetReviewSnapshot(
    schemaVersion: (json['schemaVersion'] as num?)?.toInt() ?? currentSchemaVersion,
    snapshotId: json['snapshotId'] as String,
    builtAt: (json['builtAt'] as num).toInt(),
    staleAfterMs: (json['staleAfterMs'] as num).toInt(),
    cursor: (json['cursor'] as num?)?.toInt() ?? 0,
    cards: (json['cards'] as List? ?? const [])
        .cast<Map<String, dynamic>>()
        .map(WidgetReviewCard.fromJson)
        .toList(),
  );
}

/// One fully-resolved card in a [WidgetReviewSnapshot]: the front (context sentence + the target span
/// to highlight) and the back ([meaning] + [meaningStatus]). The front always shows the word — this is
/// RECOGNITION review (see the word → recall the meaning), NOT a fill-in-the-blank cloze.
class WidgetReviewCard {
  const WidgetReviewCard({
    required this.wordId,
    required this.surfaceUnit,
    required this.targetLang,
    required this.dueAt,
    required this.state,
    required this.contextText,
    required this.targetSpan,
    required this.meaning,
    required this.meaningStatus,
    this.ipa,
    this.contextMeaning,
  });

  final String wordId;
  final String surfaceUnit;
  final String targetLang;

  /// Bare IPA for the word — the primary reading's US transcription (else UK), or null when the
  /// explanation omitted it (omit-on-fail). The renderer adds the `/…/` slashes.
  final String? ipa;

  /// Due time (epoch ms). Used ONLY for the glance count — the widget never locally re-sorts or
  /// re-filters by it (the server already ordered the queue).
  final int dueAt;

  /// `due` (already-scheduled, due now) | `new` (never reviewed — surfaced within the server's daily
  /// new-card cap). Both are reviewable in the widget; the renderer doesn't branch on this today.
  final String state;

  /// The most-recent context sentence (front), or `''` for a bare card.
  final String contextText;

  /// The UTF-16 offsets of the target word within [contextText] to highlight, or null when there's no
  /// span (bare card / no highlight). UTF-16 so Swift's `String`/`NSRange` and Dart's UTF-16 indexing
  /// agree (a committed CJK fixture pins this).
  final WidgetTargetSpan? targetSpan;

  /// The back: the per-POS senses laid out one line per part of speech ([widgetMeaningText]), or null
  /// when [meaningStatus] is not [ready]. The wire key stays `meaning` (the shipped widget reads it).
  final String? meaning;

  /// Whether the back is renderable: `ready` | `unsupported` (target language off the allowlist) |
  /// `unavailable` (fetch failure / no blob). The front is reviewable regardless.
  final WidgetMeaningStatus meaningStatus;

  /// The in-sentence "Explain here" gloss for the front sentence (word-in-context + whole-sentence), or
  /// null when none was generated. Read for FREE from the already-fetched `/contexts` (NOT the metered
  /// endpoint), so the large widget back can show it like the in-app Review card. Additive optional field
  /// — an older widget build that lacks it just ignores the key, so no [WidgetReviewSnapshot] schema bump.
  final String? contextMeaning;

  Map<String, dynamic> toJson() => {
    'wordId': wordId,
    'surfaceUnit': surfaceUnit,
    'targetLang': targetLang,
    'dueAt': dueAt,
    'state': state,
    'contextText': contextText,
    'targetSpan': targetSpan?.toJson(),
    'ipa': ipa,
    'meaning': meaning,
    'meaningStatus': meaningStatus.wire,
    'contextMeaning': contextMeaning,
  };

  factory WidgetReviewCard.fromJson(Map<String, dynamic> json) => WidgetReviewCard(
    wordId: json['wordId'] as String,
    surfaceUnit: json['surfaceUnit'] as String,
    targetLang: json['targetLang'] as String,
    dueAt: (json['dueAt'] as num).toInt(),
    state: json['state'] as String? ?? 'due',
    contextText: json['contextText'] as String? ?? '',
    targetSpan: WidgetTargetSpan.fromJson(json['targetSpan']),
    ipa: json['ipa'] as String?,
    meaning: json['meaning'] as String?,
    meaningStatus: WidgetMeaningStatus.fromWire(json['meaningStatus']),
    contextMeaning: json['contextMeaning'] as String?,
  );
}

/// A half-open UTF-16 `[start, end)` highlight range within a card's context text. Serialized as a
/// 2-element array `[start, end]` (matching the RFC), or null.
class WidgetTargetSpan {
  const WidgetTargetSpan(this.start, this.end);

  final int start;
  final int end;

  List<int> toJson() => [start, end];

  /// Build a span from raw start/end bounds, or null if either is absent or the pair is malformed
  /// (negative / inverted). Construct spans through this — NOT the raw constructor — so a
  /// non-normalized source can't put an invalid span into a snapshot (which the Swift side would then
  /// silently drop to null, an asymmetric, hard-to-debug mismatch).
  static WidgetTargetSpan? fromBounds(int? start, int? end) {
    if (start == null || end == null || start < 0 || end < start) return null;
    return WidgetTargetSpan(start, end);
  }

  /// Decode `[start, end]` defensively: a malformed / one-sided / inverted span → null (the widget
  /// then renders plain text), never a crash — mirroring the backend's span normalization.
  static WidgetTargetSpan? fromJson(Object? raw) {
    if (raw is! List || raw.length != 2) return null;
    return fromBounds((raw[0] as num?)?.toInt(), (raw[1] as num?)?.toInt());
  }

  @override
  bool operator ==(Object other) =>
      other is WidgetTargetSpan && other.start == start && other.end == end;

  @override
  int get hashCode => Object.hash(start, end);
}

/// The serialized form of a widget card's meaning state — [MeaningStatus] minus `loading` (a snapshot
/// card is always fully resolved; the widget never fetches). A stable wire string the SwiftUI side
/// switches on.
enum WidgetMeaningStatus {
  ready('ready'),
  unsupported('unsupported'),
  unavailable('unavailable');

  const WidgetMeaningStatus(this.wire);
  final String wire;

  /// Lenient decode: an unknown/absent value → [unavailable] (the safe "no meaning, still reviewable"
  /// state), so a future status string can't crash an older widget.
  static WidgetMeaningStatus fromWire(Object? raw) {
    for (final s in values) {
      if (s.wire == raw) return s;
    }
    return WidgetMeaningStatus.unavailable;
  }

  /// Project a resolved [MeaningStatus] onto the wire enum (`loading` shouldn't reach a built snapshot;
  /// it degrades to [unavailable]).
  static WidgetMeaningStatus fromMeaningStatus(MeaningStatus status) => switch (status) {
    MeaningStatus.ready => WidgetMeaningStatus.ready,
    MeaningStatus.unsupported => WidgetMeaningStatus.unsupported,
    MeaningStatus.unavailable || MeaningStatus.loading => WidgetMeaningStatus.unavailable,
  };
}
